third_party/move/move-binary-format/src/file_format.rs

```diff
@@ -2934,6 +2969,134 @@ pub enum Bytecode {
     "#]
     VecSwap(SignatureIndex),
 
+    #[group = "closure"]
+    #[description = "Load a function value onto the stack."]
+    #[static_operands = "[func_handle_idx]"]
+    #[semantics = "stack << functions[function_handle_idx]"]
+    #[runtime_check_epilogue = "ty_stack << func_ty"]
+    #[gas_type_creation_tier_1 = "func_ty"]
+    LdFunction(FunctionHandleIndex),
+
+    #[group = "closure"]
+    #[description = "Generic version of `LdFunction`."]
+    #[static_operands = "[func_inst_idx]"]
+    #[semantics = "See `LdFunction`."]
+    #[runtime_check_epilogue = "See `LdFunction`."]
+    #[gas_type_creation_tier_0 = "ty_args"]
+    #[gas_type_creation_tier_1 = "local_tys"]
+    LdFunctionGeneric(FunctionInstantiationIndex),
+
+    #[group = "closure"]
+    #[description = r#"
+        `EarlyBindFunction(|t1..tn|r with a, count)` creates new function value based
+        on the function value at top of stack by adding `count` arguments
+        popped from the stack to the closure found on top of stack.
+
+        If the function value's type has at least `count` parameters with types
+        that match the `count` arguments on the stack, then a function closure
+        capturing those values with the provided function is pushed on top of
+        stack.
+
+        Notice that the type as part of this instruction is redundant for
+        execution semantics. Since the closure is expected to be on top of the stack,
+        it can decode the arguments underneath without type information.
+        However, the type is required for the current implementation of
+        static bytecode verification.
+    "#]
+    #[static_operands = "[u8_value]"]
+    #[semantics = r#"
+        stack >> function_handle
+
+        let [func, k, [arg_0, .., arg_{k-1}]] = function_handle
+        // Information like the function signature are loaded from the file format
+        i = u8_value
+        n = func.num_params
+        if i + k > n then abort
+
+        stack >> arg'_{i-1}
+        ..
+        stack >> arg'_0
+        new_function_handle = [func, k + i, [arg_0, .., arg_{k-1}, arg'_0, .., arg'_{i-1}]]
+        stack << new_function_handle
+    "#]
+    #[runtime_check_epilogue = r#"
+        // NOT: assert func visibility rules
+        func param types match provided parameter types
+    "#]
+    #[gas_type_creation_tier_0 = "closure_ty"]
+    EarlyBindFunction(SignatureIndex, u8),
```

Comment:
Any function generated  for the body of an lambda expression can order free variables before lambda parameters so that it is natural to early-bind the `k` first parameters.  If the user wants to use an existing function with parameters in the wrong order, the user can add a wrapper function as necessary.

The mask idea and the code Wolfgang and I had written to deal with more arbitrary permutations of early/late-bound parameters was very "clever" and therefore hard to review and really an unnecessary security risk.

Reply:
replied already